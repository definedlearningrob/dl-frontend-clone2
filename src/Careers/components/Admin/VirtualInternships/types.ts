import { TOpportunity as TBaseOpportunity } from '@dc/resources/types';

export const enum VirtualInternshipStatuses {
  DRAFT = 'DRAFT',
  PUBLISHED = 'PUBLISHED',
}

export type VirtualInternshipLessonAttributes = {
  lessonId: string;
  step: number;
};

export type FormValues = {
  badges: {
    id: string;
    name: string;
    imageUrl: string;
  }[];
  description: string;
  name: string;
  pathways: {
    label: string;
    value: string;
  }[];
  availableSpots: number;
  creditsOutcomes: string;
  image: Record<string, unknown>;
  tags: {
    label: string;
    value: string;
  }[];
  requiredExperiences: number;
  status: { value: VirtualInternshipStatuses; label: string };
  calendarLessons: Lesson[];
  experienceOpportunityLessons: Lesson[];
  postExperienceLessons: Lesson[];
  readinessSkillsLessons: Lesson[];
};

export type TPathway = Pick<TBaseOpportunity['pathways'][number], 'id' | 'name'>;

export type TOpportunity = Pick<
  TBaseOpportunity,
  | 'id'
  | 'name'
  | 'availableSpots'
  | 'creditsOutcomes'
  | 'description'
  | 'imageUrl'
  | 'opportunityType'
  | 'tags'
  | 'salaryInformation'
> & {
  pathways: TPathway[];
};

export type Lesson = {
  id: string;
  step: number;
  name: string;
  imageUrl: string;
  thumbnailUrl: string;
  type: string;
};

export type TVirtualInternship = {
  archivedAt: string | null;
  badges: {
    id: string;
    name: string;
    imageUrl: string;
  }[];
  id: string;
  opportunity: TOpportunity;
  requiredExperiences: number;
  status: VirtualInternshipStatuses;
  calendarLessons: Lesson[];
  experienceOpportunityLessons: Lesson[];
  postExperienceLessons: Lesson[];
  readinessSkillsLessons: Lesson[];
};
