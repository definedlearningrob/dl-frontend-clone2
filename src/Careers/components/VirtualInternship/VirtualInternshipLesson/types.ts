import { TLesson as TStudentLesson } from '@dc/graphql/student/queries/virtualInternshipLesson';
import { TLesson as TUserLesson } from '@dc/graphql/user/queries/virtualInternshipLesson';

export type TLesson = TStudentLesson | TUserLesson;
export type TCheckInQuestion = TLesson['checkInQuestions'][number];
export type TCheckInGroup = TLesson['checkInGroups'][number];
