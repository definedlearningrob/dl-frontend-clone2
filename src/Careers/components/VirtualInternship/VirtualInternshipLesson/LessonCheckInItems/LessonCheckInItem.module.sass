@import '@shared/stylesheets/config.sass'

.checkInItemName
  font-size: $font-size-sm
  margin-bottom: $m-base

  +media-breakpoint-up(xxxl)
    font-size: $font-size-base

.questionsWrapper
  +flex-container(column)
  gap: $m-base

  +media-breakpoint-up(xxxl)
    gap: $m-md

.textArea
  background-color: $color-white
  border: 1px solid $color-border-disabled
  border-radius: $border-radius-sm
  padding: $m-x
  resize: none
  width: 100%

.checkInActions
  +flex-container(row, flex-end, center)
  margin-top: $m-base
