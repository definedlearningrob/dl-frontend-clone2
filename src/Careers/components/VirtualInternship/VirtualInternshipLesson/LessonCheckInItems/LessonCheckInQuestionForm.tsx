import { Form, useFormikContext } from 'formik';
import { useTranslation } from 'react-i18next';

import { AutoSizeTextarea } from '@shared/components/AutoSizeTextarea/AutoSizeTextarea';
import SharedButton from '@shared/components/Button/Button';
import { useRole } from '@shared/hooks/useRole';

import { TCheckInQuestion } from '../types';

type FormikValues = {
  answer?: string;
};

type Props = {
  onEditClick: () => void;
  isEditingAnswer: boolean;
  checkInQuestion: TCheckInQuestion;
};

export const LessonCheckInQuestionForm = ({
  checkInQuestion,
  isEditingAnswer,
  onEditClick,
}: Props) => {
  const { t } = useTranslation();

  const { isPublic, isUser } = useRole();
  const disabledButton = isUser || isPublic;
  const { isSubmitting, submitForm } = useFormikContext<FormikValues>();

  const disabledTextarea = !isEditingAnswer || isUser || isPublic;
  const shouldDisplaySaveButton = isEditingAnswer || isUser || isPublic;

  return (
    <Form>
      <AutoSizeTextarea
        disabled={disabledTextarea}
        labelText={checkInQuestion.question}
        name='answer'
      />
      <SharedButton
        className='ms-auto'
        disabled={disabledButton}
        isLoading={isSubmitting}
        minWidth='md'
        size='md'
        type='button'
        variant={shouldDisplaySaveButton ? 'primary' : 'primary-outlined'}
        onClick={shouldDisplaySaveButton ? submitForm : onEditClick}>
        {shouldDisplaySaveButton ? t('sharedCommon.save') : t('sharedCommon.edit')}
      </SharedButton>
    </Form>
  );
};
