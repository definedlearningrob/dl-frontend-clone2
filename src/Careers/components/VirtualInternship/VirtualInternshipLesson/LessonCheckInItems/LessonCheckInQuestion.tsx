import { useTranslation } from 'react-i18next';
import { Formik } from 'formik';
import * as Yup from 'yup';
import { GraphQLError } from 'graphql';
import { useState } from 'react';
import { isEmpty } from 'lodash-es';
import { useParams } from 'react-router-dom';

import { useSubmitCheckInQuestionAnswer } from '@dc/graphql/student/hooks/useSubmitCheckInQuestionsAnswer';
import { VIRTUAL_INTERNSHIP_CONTENT_QUERY } from '@dc/graphql/student/queries/virtualInternshipContent';

import { getFormErrors } from '@shared/utils/graphql';

import { TCheckInQuestion } from '../types';

import { LessonCheckInQuestionForm } from './LessonCheckInQuestionForm';

type Props = {
  checkInQuestion: TCheckInQuestion;
};

type FormValues = {
  id?: string;
  answer: string;
};

export const LessonCheckInQuestion = ({ checkInQuestion }: Props) => {
  const { t } = useTranslation();
  const { opportunityId } = useParams<{ opportunityId: string }>();

  const hasAnswer = 'answer' in checkInQuestion;

  const [isEditingAnswer, setIsEditingAnswer] = useState(
    hasAnswer && isEmpty(checkInQuestion.answer)
  );

  const initialValues =
    hasAnswer && checkInQuestion.answer
      ? {
          id: checkInQuestion.answer.id,
          answer: checkInQuestion.answer.answer,
        }
      : { answer: '' };

  const validationSchema = Yup.object().shape({
    answer: Yup.string().trim().required(t('validation.messages.required')),
  });

  const [submitCheckInAnswer] = useSubmitCheckInQuestionAnswer({
    checkInQuestionId: checkInQuestion.id,
  });

  const handleOnSubmit = async (values: FormValues, actions: { setErrors?: Function }) => {
    const { setErrors } = actions;

    setIsEditingAnswer(false);

    if (values.answer === initialValues.answer) {
      return;
    }

    try {
      await submitCheckInAnswer({
        values,
        refetchQueries: [{ query: VIRTUAL_INTERNSHIP_CONTENT_QUERY, variables: { opportunityId } }],
      });
    } catch (error) {
      const errors = getFormErrors(error as { graphQLErrors: GraphQLError[] });
      setErrors && setErrors(errors);
    }
  };

  const handleEditClick = () => {
    if (!isEditingAnswer) {
      setIsEditingAnswer(true);
    }
  };

  return (
    <Formik
      enableReinitialize={true}
      initialValues={initialValues}
      validationSchema={validationSchema}
      onSubmit={(values, actions) => handleOnSubmit(values, actions)}>
      <LessonCheckInQuestionForm
        checkInQuestion={checkInQuestion}
        isEditingAnswer={isEditingAnswer}
        onEditClick={handleEditClick}
      />
    </Formik>
  );
};
