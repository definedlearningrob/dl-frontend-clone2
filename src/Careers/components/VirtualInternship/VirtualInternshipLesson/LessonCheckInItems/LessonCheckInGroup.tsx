import { TCheckInGroup } from '../types';

import { LessonCheckInQuestion } from './LessonCheckInQuestion';
import styles from './LessonCheckInItem.module.sass';

type Props = {
  checkInGroup: TCheckInGroup;
};

export const LessonCheckInGroup = ({ checkInGroup }: Props) => (
  <>
    <h5 className={styles.checkInItemName}>{checkInGroup.displayName}</h5>
    <div className={styles.questionsWrapper}>
      {checkInGroup.questions.map((question) => (
        <LessonCheckInQuestion key={question.id} checkInQuestion={question} />
      ))}
    </div>
  </>
);
