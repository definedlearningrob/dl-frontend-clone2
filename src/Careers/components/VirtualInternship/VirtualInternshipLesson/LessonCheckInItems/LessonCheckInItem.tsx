import { TCheckInGroup, TCheckInQuestion } from '../types';

import { LessonCheckInGroup } from './LessonCheckInGroup';
import { LessonCheckInQuestion } from './LessonCheckInQuestion';

type Props = {
  checkInItem: TCheckInGroup | TCheckInQuestion;
};

export const LessonCheckInItem = ({ checkInItem }: Props) => {
  if (checkInItem.__typename === 'CheckInGroup') {
    return <LessonCheckInGroup checkInGroup={checkInItem} />;
  }

  return <LessonCheckInQuestion checkInQuestion={checkInItem} />;
};
