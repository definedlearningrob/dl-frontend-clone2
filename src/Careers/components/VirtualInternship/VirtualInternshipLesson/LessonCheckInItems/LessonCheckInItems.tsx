import { LessonItemCard } from '@dc/components/Student/Lesson/shared/LessonItemCard/LessonItemCard';

import { TCheckInGroup, TCheckInQuestion } from '../types';

import { LessonCheckInItem } from './LessonCheckInItem';

type Props = {
  checkInItems: (TCheckInGroup | TCheckInQuestion)[];
};

export const LessonCheckInItems = ({ checkInItems }: Props) => (
  <>
    {checkInItems.map((checkInItem) => (
      <LessonItemCard key={checkInItem.id} id={`${checkInItem.id}-${checkInItem.__typename}`}>
        <LessonCheckInItem checkInItem={checkInItem} />
      </LessonItemCard>
    ))}
  </>
);
