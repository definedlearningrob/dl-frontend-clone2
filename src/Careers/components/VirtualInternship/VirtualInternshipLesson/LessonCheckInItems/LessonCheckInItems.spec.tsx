import { MockedProvider } from '@apollo/client/testing';
import { screen, waitFor, within } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Route } from 'react-router-dom';

import { LessonCheckInItems } from '@dc/components/VirtualInternship/VirtualInternshipLesson/LessonCheckInItems/LessonCheckInItems';
import { renderWithRouterAndReduxProvider } from '@dc/utils/test';
import { CREATE_CHECKIN_QUESTION_ANSWER } from '@dc/graphql/student/mutations/createCheckInQuestionAnswer';
import { UPDATE_CHECKIN_QUESTION_ANSWER } from '@dc/graphql/student/mutations/updateCheckInQuestionAnswer';
import cacheConfig from '@dc/graphql/cacheConfig';

import { TCheckInQuestionAnswer } from '@shared/components/CheckIns/types';

type Answer = Pick<TCheckInQuestionAnswer, 'id' | 'answer'>;

const mockedCheckInItems = (answers: Answer[] = new Array(3).fill(null)) => [
  {
    __typename: 'CheckInGroup',
    displayName: 'Lesson 1 Check-ins',
    id: '3',
    questions: [
      {
        __typename: 'CheckInQuestion',
        answer: answers[0],
        id: '5',
        question: 'Whats your name?',
        step: 1,
      },
      {
        __typename: 'CheckInQuestion',
        answer: answers[1],
        id: '6',
        question: 'How are you?',
        step: 2,
      },
    ],
    step: 1,
  },
  {
    __typename: 'CheckInQuestion',
    answer: answers[2],
    id: '4',
    question: 'Whats your question?',
    step: 1,
  },
];

const lessonId = 864;

const createCheckInSpy = jest.fn();
const updateCheckInSpy = jest.fn();

const mocks = [
  {
    request: {
      query: CREATE_CHECKIN_QUESTION_ANSWER,
      variables: {
        input: {
          answer: 'John Doe',
          lessonId: `${lessonId}`,
          checkInQuestionId: '5',
        },
      },
    },
    result: () => {
      createCheckInSpy();

      return {
        data: {
          createCheckInQuestionAnswer: {
            checkInQuestionAnswer: {
              answer: 'John Doe',
              id: 'temp',
              __typename: 'CheckInQuestionAnswer',
            },
          },
          __typename: 'CreateCheckInQuestionAnswerMutationPayload',
        },
      };
    },
  },

  {
    request: {
      query: UPDATE_CHECKIN_QUESTION_ANSWER,
      variables: {
        input: {
          answer: 'Adam Snow and Tom Onion',
          id: '9999',
        },
      },
    },
    result: () => {
      updateCheckInSpy();

      return {
        data: {
          updateCheckInQuestionAnswer: {
            checkInQuestionAnswer: {
              answer: 'Adam Snow and Tom Onion',
              id: 'temp',
              __typename: 'CheckInQuestionAnswer',
            },
          },
          __typename: 'CreateCheckInQuestionAnswerMutationPayload',
        },
      };
    },
  },
];

// @ts-ignore
const renderCheckInItems = (mocks, mockedCheckInItems, userType?: 'user' | 'student') => {
  let initialState;

  if (userType) {
    initialState = {
      session: {
        user: {
          type: userType,
        },
        loginError: {},
      },
    };
  }

  return renderWithRouterAndReduxProvider(
    <Route path='/opportunities/:opportunityId/virtual-internship/lesson/:lessonId'>
      <MockedProvider cache={cacheConfig} mocks={mocks}>
        <LessonCheckInItems checkInItems={mockedCheckInItems} />
      </MockedProvider>
    </Route>,
    {
      route: `/opportunities/6/virtual-internship/lesson/${lessonId}`,
      initialState,
    }
  );
};

describe('LessonCheckInItems', () => {
  it('should render properly', () => {
    const { container } = renderCheckInItems(mocks, mockedCheckInItems(), 'student');

    expect(container).toHaveTextContent('Lesson 1 Check-ins');

    expect(container).toMatchSnapshot();
  });

  it(`when user is logged in - input should be enabled`, () => {
    renderCheckInItems(mocks, mockedCheckInItems(), 'student');

    const firstQuestion = screen.getByLabelText('Whats your name?');

    expect(firstQuestion).toBeEnabled();
  });

  it(`when user it not logged in - input should be disabled`, () => {
    renderCheckInItems(mocks, mockedCheckInItems());

    const firstQuestion = screen.getByLabelText('Whats your name?');

    expect(firstQuestion).toBeDisabled();
  });

  it('should add checkin answer', async () => {
    renderCheckInItems(mocks, mockedCheckInItems(), 'student');

    const firstQuestionForm = screen.getByLabelText('Whats your name?').closest('form');

    const firstQuestion = screen.getByLabelText('Whats your name?');

    userEvent.click(firstQuestion);

    userEvent.paste(firstQuestion, 'John Doe');

    expect(firstQuestion).toHaveValue('John Doe');

    const saveButton = within(firstQuestionForm!).getByRole('button', { name: 'Save' });

    expect(createCheckInSpy).toBeCalledTimes(0);

    userEvent.click(saveButton);

    await waitFor(() => {
      expect(createCheckInSpy).toBeCalledTimes(1);
    });
  });

  it('should edit check-in answer', async () => {
    const answer = {
      answer: 'Adam Snow',
      checkInQuestionId: '5',
      id: '9999',
      studentId: '1234',
      grade: null,
      __typename: 'CheckInQuestionAnswer',
    };

    renderCheckInItems(mocks, mockedCheckInItems([answer]), 'student');

    const firstQuestionForm = screen.getByLabelText('Whats your name?').closest('form');
    const firstQuestion = screen.getByLabelText('Whats your name?');

    const editButton = within(firstQuestionForm!).getByRole('button', { name: 'Edit' });

    expect(firstQuestion).toBeDisabled();
    expect(firstQuestion).toHaveValue('Adam Snow');

    userEvent.click(editButton);

    expect(firstQuestion).toBeEnabled();

    userEvent.paste(firstQuestion, ' and Tom Onion');

    expect(firstQuestion).toHaveValue('Adam Snow and Tom Onion');

    const saveButton = within(firstQuestionForm!).getByRole('button', { name: 'Save' });

    expect(updateCheckInSpy).toHaveBeenCalledTimes(0);

    userEvent.click(saveButton);

    const updatedEditButton = await within(firstQuestionForm!).findByRole('button', {
      name: 'Edit',
    });

    expect(updatedEditButton).toBeInTheDocument();

    await waitFor(() => {
      expect(updateCheckInSpy).toHaveBeenCalledTimes(1);
    });

    expect(firstQuestion).toBeDisabled();
  });
});
