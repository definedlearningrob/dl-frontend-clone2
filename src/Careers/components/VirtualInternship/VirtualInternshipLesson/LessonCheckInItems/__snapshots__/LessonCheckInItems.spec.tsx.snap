// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LessonCheckInItems should render properly 1`] = `
<div>
  <div
    class="card lessonCard"
    data-testid="lesson-element-card"
    id="3-CheckInGroup"
  >
    <h5
      class="checkInItemName"
    >
      Lesson 1 Check-ins
    </h5>
    <div
      class="questionsWrapper"
    >
      <form
        action="#"
      >
        <div
          class="textarea"
        >
          <label
            class="w-full"
          >
            <div
              class="text-sm xxxl:text-base mb-sm font-bold"
            >
              Whats your name?
            </div>
            <textarea
              class="scrollbar textareaInput"
              name="answer"
              placeholder="Type here..."
            />
          </label>
          <div
            class="inputErrorMessage"
          />
        </div>
        <button
          class="button flex items-center justify-center gap-xs rounded-sm cursor-pointer text-sm font-medium no-underline focus-visible:outline focus-visible:outline-1 focus-visible:outline-offset-[3px] ms-auto -primary bg-primary-500 hover:bg-primary-600 outline-primary-500 border-primary-500 py-xs px-base leading-[22px] -min-width-md border text-white"
          data-testid="button"
          type="button"
        >
          <span
            class="items-center flex justify-center gap-xs"
          >
            Save
          </span>
        </button>
      </form>
      <form
        action="#"
      >
        <div
          class="textarea"
        >
          <label
            class="w-full"
          >
            <div
              class="text-sm xxxl:text-base mb-sm font-bold"
            >
              How are you?
            </div>
            <textarea
              class="scrollbar textareaInput"
              name="answer"
              placeholder="Type here..."
            />
          </label>
          <div
            class="inputErrorMessage"
          />
        </div>
        <button
          class="button flex items-center justify-center gap-xs rounded-sm cursor-pointer text-sm font-medium no-underline focus-visible:outline focus-visible:outline-1 focus-visible:outline-offset-[3px] ms-auto -primary bg-primary-500 hover:bg-primary-600 outline-primary-500 border-primary-500 py-xs px-base leading-[22px] -min-width-md border text-white"
          data-testid="button"
          type="button"
        >
          <span
            class="items-center flex justify-center gap-xs"
          >
            Save
          </span>
        </button>
      </form>
    </div>
  </div>
  <div
    class="card lessonCard"
    data-testid="lesson-element-card"
    id="4-CheckInQuestion"
  >
    <form
      action="#"
    >
      <div
        class="textarea"
      >
        <label
          class="w-full"
        >
          <div
            class="text-sm xxxl:text-base mb-sm font-bold"
          >
            Whats your question?
          </div>
          <textarea
            class="scrollbar textareaInput"
            name="answer"
            placeholder="Type here..."
          />
        </label>
        <div
          class="inputErrorMessage"
        />
      </div>
      <button
        class="button flex items-center justify-center gap-xs rounded-sm cursor-pointer text-sm font-medium no-underline focus-visible:outline focus-visible:outline-1 focus-visible:outline-offset-[3px] ms-auto -primary bg-primary-500 hover:bg-primary-600 outline-primary-500 border-primary-500 py-xs px-base leading-[22px] -min-width-md border text-white"
        data-testid="button"
        type="button"
      >
        <span
          class="items-center flex justify-center gap-xs"
        >
          Save
        </span>
      </button>
    </form>
  </div>
</div>
`;
